### PR TITLE
add new plugin : focalfp

### PR DESCRIPTION
**Pull request by waynehuang2022**
Type of pull request:

- [X] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by hughsie** [comment]
I'll rebase this and also fix up some of the easy fix stuff.

**Comment by hughsie** [comment]
@waynehuang2022 I've fixed up some of the design issues in the plugin, but I'm sure I've broken something or used G_LITTLE_ENDIAN when I should have used G_BIG_ENDIAN or something like that. Can you try my version of the plugin again and fix up any of my mistakes? You can just push "fixup" commits to that branch to update this PR.

File: plugins/focalfp/focalfp.quirk [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[HIDRAW\VEN_2808&DEV_0106]
```
Comment (superm1) [inline plugins/focalfp/focalfp.quirk:1]: could this apply to all `VEN 2808`?

File: plugins/focalfp/fu-focalfp-firmware.c [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	/* start address */
+	if (!fu_memread_uint16_safe(buf,
+				    bufsz,
+				    FOCAL_NAME_START_ADDR_WRDS,
+				    &self->start_address,
+				    G_BIG_ENDIAN,
+				    error)) {
+		return FALSE;
+	}
+	if (self->start_address != 0x582e) {
```
Comment (superm1) [inline plugins/focalfp/fu-focalfp-firmware.c:61]: is this going to be the same for all devices?

File: plugins/focalfp/fu-focalfp-hid-device.c [hunk 3]
```diff
@@ -0,0 +1,662 @@
...
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "is not correct subsystem=%s, expected hidraw",
+			    fu_udev_device_get_subsystem(FU_UDEV_DEVICE(device)));
+		return FALSE;
+	}
+
+	/* i2c-hid */
+	if (fu_udev_device_get_model(FU_UDEV_DEVICE(device)) != 0x0106) {
```
Comment (superm1) [inline plugins/focalfp/fu-focalfp-hid-device.c:63]: will these all be the same every time?

File: plugins/focalfp/focalfp.quirk (hunk 1 above)
Comment (hughsie) [reply plugins/focalfp/focalfp.quirk:1]: I actually asked them to limit this for a first take; we don't know how well all the touchpads are going to respond to probing yet.

File: plugins/focalfp/fu-focalfp-hid-device.c (hunk 3 above)
Comment (waynehuang2022) [reply plugins/focalfp/fu-focalfp-hid-device.c:63]: Yes, these will be the same every time.

File: plugins/focalfp/fu-focalfp-firmware.c (hunk 2 above)
Comment (waynehuang2022) [reply plugins/focalfp/fu-focalfp-firmware.c:61]: Yes.

**Comment by hughsie** [comment]
@waynehuang2022 can you check this still works on actual hardware please? I can't believe I didn't break *something* :)

**Comment by waynehuang2022** [comment]
@hughsie  I had checked.I will push commits after complete the modification.

**Comment by hughsie** [comment]
@waynehuang2022 I've rebased and squashed down all the fixes and then with your approval on the PR we can merge this!

**Review by waynehuang2022 — COMMENTED** [review]
approved

**Review by waynehuang2022 — COMMENTED** [review]
approved

**Comment by waynehuang2022** [comment]
@hughsie Thank you for your help during this time .

**Comment by hughsie** [comment]
No worries; any further changes to the plugin please create a new branch, open a new PR and we can review and merge in the same way.

**Pull request merged**